### ISSUE TITLE
Consider changing duk_(u)int_t to (u)intptr_t at least on Windows

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#1241]
On Windows int and unsigned int are 32-bit even on 64-bit platforms. It may be less surprising if `duk_(u)int_t` were 64-bit.

**Comment by fatcerberus** [comment]
Isn't `int` 32-bit also on Linux?  The surprising thing on Windows is that `long` is *also* 32-bit (!).

**Comment by svaarala** [comment]
You're right, it's a more generic issue, changed the title and description. The question came from a Windows use case where `duk_push_int()` would truncate a 64-bit value which seemed surprising.

Older Lua versions used `ptrdiff_t` as the base integer type if I recall correctly. Now it seems to provide several alternatives: https://github.com/lua/lua/blob/master/luaconf.h#L540-L590.

**Comment by fatcerberus** [comment]
Thinking about this, I think it might be necessary going forward for Duktape to use a 64-bit type as the base integer type anyway.  Many operations in ES6+ dealing with integers specify a range of (-2**53,+2**53), i.e. the IEEE double integer range, which is impossible to represent with only 32 bits.

The downside to using 64-bit types across the board is that on some platforms, e.g. x86, operations on 64-bit integers are not atomic and thus might be a performance issue.

**Comment by svaarala** [comment]
Also some platforms don't have a 64-bit type at all, so it can't be a mandatory requirement.

**Comment by svaarala** [comment]
More broadly, coming up with a good C typing isn't actually trivial. A simple approach is to just use a single integer type (with maybe a signed and unsigned distinction which is often practical). But the problem with that is that the type may need to be larger than ideal (as you pointed out). Using a 64-bit type on a platform with 32-bit registers has an impact on both footprint and performance which isn't negligible.

Another approach, which the API tries mostly to follow now, is to have a few integer typedefs for different purposes. For example, `duk_idx_t` is for value stack indices and can be limited to 32 bits because going beyond >32GB value stacks is not really that important. I also added a `duk_uarridx_t` to see how that works in practice, anticipating that the array index might go beyond 32 bits at some point (so far it hasn't though, but string and buffer indices could go beyond 32 bits). The downsides of this is that an application writer may need to think about type sizes and what types can be assigned without explicit casts and such. Using a `duk_idx_t` as a loop variable, for example, can be seen as awkward.

So maybe those places where a 64-bit integer could be needed have some common theme (say an "integer index" type?) and can get its own typedef and be used in the APIs where it matters.

But even so, on some platforms the index may need to be 32-bit because there just isn't support for 64-bit values or support exists but is too broken to be usable (I'm working with one such platform right now...). In practice this is not very limiting, because you're unlikely to be working with strings or buffers beyond 4G elements on such platforms, for example. It's possible to have sparse structures that large (say a sparse array, if arrays get >4G index support) but it's still not a very common case.

**Comment by svaarala** [comment]
Just as a footprint ballpark, on x86 changing `duk_(u)int_t` to `(u)int64_t` adds slightly over 40kB (!) of footprint. So while a 64-bit type would be fine for, say, a "get codepoint at string index X" API call, as a general type it'd be quite awkward.

**Comment by svaarala** [comment]
A few approaches not maybe covered above:

- Keep `duk_(u)int_t` as a usually 32-bit type, but ensure that each API call taking an "integer index" has a variant also accepting `duk_double_t` which covers the full 53-bit range. Not ideal, but doesn't impact portability and provides access to the full range.
- Keep `duk_(u)int_t` as a usually 32-bit type. Add a separate large integer type which is 64 bits or more, and add API call variants which use that type for all "integer index" related API primitives. On platforms without 64-bit types those API calls wouldn't be available and wouldn't link.

When I'm working with a 32-bit platform, I would actually prefer that the API calls dealing with integer indices would be confined to 32 bits even when the technical range is larger: that added range is almost never needed in practice. So some sort of hybrid API where the full range is accessible via IEEE double indices but otherwise one is confined to the native integer would be quite workable in these cases.

For 64-bit platforms that's not necessarily the case because the cost of a 64-bit integer type is small. Changing duk_(u)int_t to a 64-bit type would be feasible there. Testing on Linux x64, the footprint impact is around +4kB which is much smaller (but still significant).

**Comment by fatcerberus** [comment]
That footprint impact is impressive, but doesn't actually surprise me too much.  On x86 operations on 64-bit integers are not atomic so even something as simple as addition may require multiple instructions.  That adds up quickly. :)

For my part I think the most reasonable way forward would be to use the native word size for `duk_[u]int_t` and then perhaps allow the user to customize with a config option if needed.

**Comment by svaarala** [comment]
That would be a relatively simple step. Although types are not currently affected by config options, it wouldn't be difficult to add options to `configure.py` to control the type selection. Except for that it's already possible (if messy) to edit `duk_config.h` and tweak the types manually.

One concrete case where some typing rework will come up is adding support for >4G buffers. For them using `duk_size_t` would probably be quite appropriate. Same thing with string integer indices, as neither are likely to be virtualized.

Longer term the C types will still need some work though. Maybe digging deeper into ES6 will provide more insight what the best solutions are. (One small but useful change would be to remove the internal distinction between `duk_context *` and `duk_hthread *` which causes a lot of unnecessary casting in the internals. I've had a pending branch for this for a while now but it depends on one more header type visibility change.)

Unfortunately even if >32-bit indices cannot map to concrete values, the internals may need to at least recognize 53-bit integer indices above the integer type range. Doing that without a 64-bit integer type is a bit awkward, but it's possible, for example, to detect and deal with the 53-bit values above the integer type range even without actually getting the integer index value back.
